ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java

```diff
@@ -269,7 +272,20 @@ public static Handler requireAuth(Class<? extends Client<?, ?>> clientType, Auth
    */
   public static Promise<UserProfile> login(Context ctx, Class<? extends Client<?, ?>> clientType) {
     return userProfile(ctx)
-      .route(p -> !p.isPresent(), p -> initiateAuthentication(ctx, clientType))
+      .flatMap(p -> {
+        if (!p.isPresent() && !supportsRedirect(clientType)) {
+          return performDirectAuthentication(ctx, clientType);
+        } else {
+          return Promise.value(p);
+        }
+      })
+      .route(p -> !p.isPresent(), p -> {
+        if (supportsRedirect(clientType)) {
+          initiateAuthentication(ctx, clientType);
+        } else {
+          ctx.clientError(401);
+        }
+      })
```

Comment:
The `flatMap` attempts to get credentials & user profile via direct auth if possible. If we're not doing direct auth or couldn't get valid credentials then we flow into the `route` which will either do redirecting auth as before or just return 401 for authentication clients that don't support redirects.